lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php

```diff
@@ -1414,6 +1420,53 @@ protected function _validateAndCompleteFieldMapping(array &$mapping)
             throw MappingException::missingFieldName($this->name);
         }
 
+        if (
+            PHP_VERSION_ID >= 70400
+            && isset($this->reflClass)
+            && $this->reflClass->hasProperty($mapping['fieldName'])
+        ) {
+            $property = $this->reflClass->getProperty($mapping['fieldName']);
+            $type     = $property->getType();
+
+            if ($type) {
+                if (! isset($mapping['nullable'])) {
+                    $mapping['nullable'] = $type->allowsNull();
+                }
+
+                if (
+                    ! isset($mapping['type'])
+                    && ($type instanceof ReflectionNamedType)
+                ) {
+                    switch ($type->getName()) {
+                        case DateInterval::class:
+                            $mapping['type'] = 'dateinterval';
+                            break;
+                        case DateTime::class:
+                            $mapping['type'] = 'datetime';
+                            break;
+                        case DateTimeImmutable::class:
+                            $mapping['type'] = 'datetime_immutable';
+                            break;
+                        case 'array':
```

Comment:
this is ambiguous, could also be `simple_array` or `object` type. Not sure what to do, since otherwise it will default to `"string"` later.

Reply:
My reasoning was to provide useful defaults instead of solving every possible use case. In this example having `"string"` as property type would simply fail at runtime as ORM would try to assign `string` value to `array` typed property. Therefore I believe useful default is better than nothing.

But the same problem comes with `\DateType` and many other fields. I believe the best long-term solution would be to create new interface, for example `DefaultTypeResolverInterface` that could look like this:
```
interface DefaultTypeResolverInterface
{
    public function determineTypeForProperty(\ReflectionType $type): ?string;
}
```
And provide at least one implementation (maybe working as here) and then allow users to choose implementation based on their config.

This way we could provide useful defaults for everyone and allow users to have simple method of providing own rules (for example I would benefit from overwriting `\DateTimeImmutable` in my projects).

I don't know how to approach such solution in ORM structure, but would love to do so. If this seems reasonable please just give me some directions and I will follow through.

Reply:
Yes, lets stay with the hardcoded resolving rules for now and we'll leave the replacable resolving rules for a future PR.

You should document what PHP types get mapped to which Doctrine types in the documentation, see `docs/` folder.

Reply:
OK, I have added documentation where it seemed logical to me